src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +128,127 @@ object RockSlicer {
     sc.stop()
   }
 
-  // Produce joints that achieve some lead balancing when generating the initial Block RDD
-  private def generateSeedJoints(joints: Seq[Joint], numSeedJoints: Integer): (Seq[Joint], Seq[Joint]) = {
-    val (persistentJoints, nonPersistentJoints) = joints.partition(_.shape.isEmpty)
-    persistentJoints.length match {
-      // All of the seed joints will be persistent
-      case len if len >= numSeedJoints => (persistentJoints.take(numSeedJoints),
-          persistentJoints.drop(numSeedJoints) ++ nonPersistentJoints)
-      // All persistent joints are used as seed joints, along with some non-persistent joints
-      case len =>
-        val numNonPersistent = numSeedJoints - len
-        // Arrange joints so we prefer elements near middle of rock mass
-        val sortedNonPersistent = nonPersistentJoints.sortWith(_.centerX < _.centerX)
-        val indices = loadBalanceIndices(sortedNonPersistent.indices).take(numNonPersistent)
-        val seedNonPersistent = indices.collect(sortedNonPersistent)
-        val remainingNonPersistent = sortedNonPersistent.diff(seedNonPersistent)
-        (persistentJoints ++ seedNonPersistent, remainingNonPersistent)
-    }
-  }
-
-  private def loadBalanceIndices(indices: Seq[Int]): Seq[Int] = {
-    def loadBalanceIndicesImpl(queue: mutable.PriorityQueue[Seq[Int]]): Seq[Int] = {
-      if (queue.isEmpty) {
-        Nil
-      } else {
-        val s = queue.dequeue()
-        val (median, left, right) = removeMedian(s)
-        if (left.nonEmpty) {
-          queue.enqueue(left)
+  /**
+    * Recursive function that merges sequence of blocks containing processor joints
+    * along matching processor joints, elliminating false blocks caused by division of 
+    * input rock volume into equal volumes by load balancer
+    * @param processorBlocks Sequence of blocks containing one or more processor joints each
+    * @param mergedBlocks Sequence of blocks that have had processor joints removed. Initial
+    *                     input is empty sequence
+    * @param origin Origin that all blocks and their faces will be referenced to for consistency
+    * @return Sequence of blocks with all processor joints removed by merging input blocks along
+    *         matching processor joints. Redundant faces and duplicate blocks are already removed
+    *         in this list, so it is not necessary to perform these computations on the returned
+    *         sequence.
+    */
+  @tailrec
+  def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
+                  origin: (Double, Double, Double)): Seq[Block] = {
+    val reconstructedBlocks = (processorBlocks map { block =>
```

Comment:
You can make this `processorBlocks.tail map ...` to avoid duplicate comparisons.